core/matrix/bccoo_helper.hpp

```diff
@@ -0,0 +1,609 @@
+                                              uint8* chunk_data,
+                                              compr_idxs<IndexType>& idxs)
+{
+    while (row_mat_data != idxs.row) {
+        idxs.row++;
+        idxs.col = 0;
+        set_value_chunk_and_increment<uint8>(chunk_data, idxs.shf,
+                                             cst_mark_end_row);
+    }
+    return (col_mat_data - idxs.col);
+}
+
+
+template <typename IndexType>
+inline void get_detect_endblock(const IndexType block_size,
+                                compr_idxs<IndexType>& idxs)
+{
+    if (idxs.nblk == block_size) {
+        idxs.nblk = 0;
+        idxs.blk++;
+    }
+}
+
+
+template <typename IndexType>
+inline void put_detect_endblock(size_type* offsets_data,
+                                const IndexType block_size,
+                                compr_idxs<IndexType>& idxs)
+{
+    if (idxs.nblk == block_size) {
+        idxs.nblk = 0;
+        idxs.blk++;
+        offsets_data[idxs.blk] = idxs.shf;
+    }
+}
+
+
+template <typename IndexType>
+inline void cnt_detect_endblock(const IndexType block_size,
+                                compr_idxs<IndexType>& idxs)
+{
+    if (idxs.nblk == block_size) {
+        idxs.nblk = 0;
+        idxs.blk++;
+    }
+}
+
+
+/*
+ *  Routines for managing block compression objects
+ */
+
+
+template <typename IndexType>
+inline void proc_block_indices(const IndexType row, const IndexType col,
+                               const compr_idxs<IndexType>& idxs,
+                               compr_blk_idxs<IndexType>& blk_idxs)
+{
+    if (idxs.nblk == 0) {
+        blk_idxs = {};
+        blk_idxs.row_frst = row;
+        blk_idxs.col_frst = col;
+    }
+    if (row != blk_idxs.row_frst) {
+        blk_idxs.rows_cols |= type_mask_rows_multiple;
+        if (row > (blk_idxs.row_frst + blk_idxs.row_diff)) {
+            blk_idxs.row_diff = row - blk_idxs.row_frst;
+        }
+    }
+    if (col < blk_idxs.col_frst) {
+        blk_idxs.col_diff += (blk_idxs.col_frst - col);
+        blk_idxs.col_frst = col;
+    } else if (col > (blk_idxs.col_frst + blk_idxs.col_diff)) {
+        blk_idxs.col_diff = col - blk_idxs.col_frst;
+    }
+}
+
+
+template <typename IndexType, typename ValueType>
+inline void cnt_block_indices(const IndexType block_size,
+                              const compr_blk_idxs<IndexType>& blk_idxs,
+                              compr_idxs<IndexType>& idxs)
+{
+    if (blk_idxs.row_diff > 0) {
+        idxs.shf += ((blk_idxs.row_diff > cst_max_size_small_idxs_row)
+                         ? sizeof(uint16)
+                         : sizeof(uint8)) *
+                    block_size;
+    }
+    if (blk_idxs.col_diff <= cst_max_size_small_row) {
+        idxs.shf += block_size * sizeof(uint8);
+    } else if (blk_idxs.col_diff <= cst_max_size_medium_row) {
+        idxs.shf += block_size * sizeof(uint16);
+    } else {
+        idxs.shf += block_size * sizeof(uint32);
```

Comment:
The problem would be if col_diff were greater than 2^32-1. It is not considered now, but it could be.
Initially, the format only considered the maximum row_diff to be 0xFF and I verified that it might be not true when blocksize is greater than 256, and now 16-bit values are considered. In the same way, I could add uint64 for col_diff if you consider